xcode/Subconscious/Shared/Services/UserProfileService.swift

```diff
@@ -344,6 +345,39 @@ actor UserProfileService {
         .eraseToAnyPublisher()
     }
     
+    func buildUserProfile(
+        address: Slashlink
+    ) async throws -> UserProfile {
+        if let petname = address.petname,
+           let cacheHit = self.cache[petname] {
+            return cacheHit
+        }
+        
+        let did = try await self.noosphere.resolve(peer: address.peer)
+        
+        let resolutionStatus: ResolutionStatus = try await Func.run {
+            guard let petname = address.petname else {
+                let sphere = try await self.noosphere.sphere(address: address)
+                let cid = try await sphere.version()
+                return .resolved(cid)
+            }
+            
+            return try await self.addressBook.resolutionStatus(petname: petname)
+        }
+        
+        let profile = try await self.loadProfileFromMemo(
+            did: did,
+            address: address,
+            resolutionStatus: resolutionStatus
+        )
+        
+        if let petname = address.petname {
+            self.cache.updateValue(profile, forKey: petname)
+        }
+        
+        return profile
+    }
```

Comment:
I needed a more general purpose way of fetching the minimal profile details and, in the process, I realised that our loading performance was gated by fetching profiles. I implemented a simple cache of petname to profile which is cleared whenever we sync, so theoretically it will always be up-to-date.